cfg/aesbackend.go

```diff
@@ -0,0 +1,226 @@
+/*
+ * knoxite
+ *     Copyright (c) 2020, Nicolas Martin <penguwin@penguwin.eu>
+ *
+ *   For license see LICENSE
+ */
+package cfg
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/knoxite/knoxite/utils"
+	"golang.org/x/crypto/scrypt"
+)
+
+// EncryptedHeaderPrefix is added to the encrypted configuration
+// to make it possible to detect it's an encrypted configuration file
+const EncryptedHeaderPrefix = "knoxiteconf+"
+
+var (
+	password string
+)
+
+// AESBackend symmetrically encrypts the configuration file using AES-GCM
+type AESBackend struct{}
+
+// NewAESBackend creates the backend.
+//
+// Given the password is required to encrypt/decrypt the configuration, if the
+// URL passed doesn't have a password or PasswordEnvVar is not defined,
+// it'll return an error.
+func NewAESBackend(u *url.URL) (*AESBackend, error) {
+	if _, err := getPassword(u); err != nil {
+		return nil, err
+	}
+
+	return &AESBackend{}, nil
+}
+
+// IsEncrypted returns true and no error if the configuration is encrypted
+//
+// If the error returned is not nil, an error was returned while opening or
+// reading the file.
+func IsEncrypted(u *url.URL) (bool, error) {
+	f, err := os.Open(u.Path)
+	if err != nil {
+		return false, err
+	}
+	defer f.Close()
+
+	b := make([]byte, 12)
+	_, err = f.Read(b)
+	if err != nil {
+		return false, err
+	}
+
+	if string(b) != EncryptedHeaderPrefix {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Load configuration file from the given URL and decrypt it
+func (b *AESBackend) Load(u *url.URL) (*Config, error) {
+	config := &Config{url: u}
+
+	if !exist(u.Path) {
+		return config, nil
+	}
+
+	ciphertext, err := ioutil.ReadFile(u.Path)
+	if err != nil {
+		return nil, err
+	}
+	ftype := ciphertext[0:12]
+	if string(ftype) != EncryptedHeaderPrefix {
+		return nil, errors.New("encrypted configuration header not valid")
+	}
+
+	p, err := getPassword(u)
+	if err != nil {
+		return nil, err
+	}
+
+	plaintext, err := decrypt(ciphertext[12:], []byte(p))
```

Comment:
Should the 12 be moved to constants? Its dependent on the length of the `EncryptedHeaderPrefix` - if the prefix changes in the future, the prefix needs to be changed in the code as well.

Reply:
Using `ciphertext[len(EncryptedHeaderPrefix):]` now